node/libs/protobuf_build/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+//! It is not possible to depend on a different proto bundle within the same crate (because
+//! these are being built simultaneously from the same build script).
+//!
+//! cargo build --all-targets
+//! perl -ne 'print "$1\n" if /PROTOBUF_DESCRIPTOR="(.*)"/' `find target/debug/build/*/output -type f` | xargs cat > /tmp/sum.binpb
+//! buf breaking /tmp/sum.binpb --against /tmp/sum.binpb
+#![allow(clippy::print_stdout)]
+use anyhow::Context as _;
+pub use once_cell::sync::Lazy;
+// Imports accessed from the generated code.
+pub use prost;
+use prost::Message as _;
+pub use prost_reflect;
+use std::{fs, path::Path, sync::Mutex};
+pub use syntax::*;
+
+mod canonical;
+mod ident;
+mod syntax;
+
+/// Traversed all the files in a directory recursively.
+fn traverse_files(
+    path: &Path,
+    f: &mut impl FnMut(&Path) -> anyhow::Result<()>,
+) -> anyhow::Result<()> {
+    if !path.is_dir() {
+        f(path).with_context(|| path.to_str().unwrap().to_string())?;
+        return Ok(());
+    }
+    for entry in fs::read_dir(path)? {
+        traverse_files(&entry?.path(), f)?;
+    }
+    Ok(())
+}
+
+/// Protobuf descriptor + info about the mapping to rust code.
+pub struct Descriptor {
+    /// Root proto package that all proto files in this descriptor belong to.
+    /// Rust types have been generated relative to this root.
+    proto_root: ProtoName,
+    /// Raw descriptor proto.
+    descriptor_proto: prost_types::FileDescriptorSet,
+    /// Direct dependencies of this descriptor.
+    dependencies: Vec<&'static Descriptor>,
+}
+
+impl Descriptor {
+    /// Constructs a Descriptor.
+    pub fn new(
+        proto_root: ProtoName,
+        dependencies: Vec<&'static Descriptor>,
+        descriptor_bytes: &impl AsRef<[u8]>,
+    ) -> Self {
+        Descriptor {
+            proto_root,
+            dependencies,
+            descriptor_proto: prost_types::FileDescriptorSet::decode(descriptor_bytes.as_ref())
+                .unwrap(),
+        }
+    }
+
+    /// Loads the descriptor to the pool, if not already loaded.
+    pub fn load(&self, pool: &mut prost_reflect::DescriptorPool) -> anyhow::Result<()> {
+        if self
+            .descriptor_proto
+            .file
+            .iter()
+            .all(|f| pool.get_file_by_name(f.name()).is_some())
+        {
+            return Ok(());
+        }
+        for d in &self.dependencies {
+            d.load(pool)?;
+        }
+        pool.add_file_descriptor_set(self.descriptor_proto.clone())?;
+        Ok(())
+    }
+
+    /// Loads the descriptor to the global pool and returns a copy of the global pool.
+    pub fn load_global(&self) -> prost_reflect::DescriptorPool {
+        /// Global descriptor pool.
+        static POOL: Lazy<Mutex<prost_reflect::DescriptorPool>> = Lazy::new(Mutex::default);
+        let pool = &mut POOL.lock().unwrap();
+        self.load(pool).unwrap();
+        pool.clone()
+    }
+}
+
+/// Code generation config. Use it in build scripts.
+pub struct Config {
```

Comment:
Bikeshedding: IMO, it could make sense to use a builder pattern here (cf. `prost_build::Config`), since manyfields have reasonable defaults. E.g., `input_root` may default to `proto`, `dependencies` to an empty `Vec`, `protobuf_crate` to `::zksync_protobuf`.

Reply:
I want the user to set all the fields and understand what they do.